### PR TITLE
Feature/NR-57 Add method specific iteration delays to runner

### PR DESCRIPTION
**Pull request by lukatarman**
Closes issue: #166

### CODE REVIEW

File: backend/src/main.js [hunk 1]
```diff
@@ -64,17 +64,20 @@ async function main() {
...
-    gameIdentifier.tryViaSteamchartsWeb,
-    gameIdentifier.updateGamesWithoutDetails,
+    { method: steamAppsAggregator.collectSteamApps },
+    { method: gameIdentifier.tryViaSteamWeb },
+    { method: gameIdentifier.tryViaSteamchartsWeb },
+    { method: gameIdentifier.updateGamesWithoutDetails },
+    { method: playerHistoryAggregator.addPlayerHistoryFromSteamcharts },
+    { method: playerHistoryAggregator.addCurrentPlayers },
     /**
      * @todo batch delay must be performed by runner
```
Comment (stas-at-ibm) [inline backend/src/main.js:74]: todo is obsolete now

File: docker-compose.yaml [hunk 2]
```diff
@@ -15,9 +15,10 @@ services:
       FEATURES_BATCH_DELAY: "${FEATURES_BATCH_DELAY}"
       FEATURES_UNIT_DELAY: "${FEATURES_UNIT_DELAY}"
       FEATURES_CURRENT_PLAYERS_UPDATE_INTERVAL_DELAY: "${FEATURES_CURRENT_PLAYERS_UPDATE_INTERVAL_DELAY}"
+      FEATURES_RELEASE_DATE_UPDATE_INTERVAL_DELAY: "${FEATURES_RELEASE_DATE_UPDATE_INTERVAL_DELAY}"
       FEATURES_UPDATE_INTERVAL_DELAY: "${FEATURES_UPDATE_INTERVAL_DELAY}"
       FEATURES_ITERATION_DELAY: "${FEATURES_ITERATION_DELAY}"
-      RUNNER_ITERATION_DELAY: "${RUNNER_ITERATION_DELAY}"
+      RUNNER_DEFAULT_ITERATION_DELAY: "${FEATURES_ITERATION_DELAY}"
```
Comment (stas-at-ibm) [inline docker-compose.yaml:21]: unused env var: RUNNER_DEFAULT_ITERATION_DELAY

Switche to RUNNER_DEFAULT_ITERATION_DELAY for the default delay. Remove FEATURES_ITERATION_DELAY. Also update the config.loader accordingly.

File: backend/src/main.js [hunk 3]
```diff
@@ -64,17 +64,20 @@ async function main() {
...
      */
-    playerHistoryAggregator.addPlayerHistoryFromSteamcharts,
-    playerHistoryAggregator.addCurrentPlayers,
   ];
-  const runner = new Runner(logger, config.runner.delayFn, config.runner.iterationDelay);
+  const runner = new Runner(logger, {
```
Comment (stas-at-ibm) [inline backend/src/main.js:77]: pass in `config.runner.options` as in master now

File: backend/src/utils/config.loader.js [hunk 4]
```diff
@@ -25,12 +25,15 @@ export const config = {
...
       Number(process.env.FEATURES_CURRENT_PLAYERS_UPDATE_INTERVAL_DELAY),
     ),
+    releaseDateUpdateIntervalDelay: Number(
+      process.env.FEATURES_RELEASE_DATE_UPDATE_INTERVAL_DELAY,
+    ),
     updateIntervalDelay: hoursToMs(Number(process.env.FEATURES_UPDATE_INTERVAL_DELAY)),
     iterationDelay: Number(process.env.FEATURES_ITERATION_DELAY),
   },
   runner: {
     delayFn: delay,
```
Comment (stas-at-ibm) [inline backend/src/utils/config.loader.js:35]: use options as in runner in master

File: backend/src/utils/runner.js [hunk 5]
```diff
@@ -1,19 +1,26 @@
...
-
-  constructor(
-    logger,
-    delayFn,
-    iterationDelay = 0,
-    iterations = Number.POSITIVE_INFINITY,
-  ) {
+  #options;
+
+  constructor(logger, options = defaultOptions()) {
```
Comment (stas-at-ibm) [inline backend/src/utils/runner.js:13]: `defaultOptions` is an object, not a function

File: docker-compose.yaml [hunk 6]
```diff
@@ -15,9 +15,10 @@ services:
       FEATURES_BATCH_DELAY: "${FEATURES_BATCH_DELAY}"
       FEATURES_UNIT_DELAY: "${FEATURES_UNIT_DELAY}"
       FEATURES_CURRENT_PLAYERS_UPDATE_INTERVAL_DELAY: "${FEATURES_CURRENT_PLAYERS_UPDATE_INTERVAL_DELAY}"
+      FEATURES_RELEASE_DATE_UPDATE_INTERVAL_DELAY: "${FEATURES_RELEASE_DATE_UPDATE_INTERVAL_DELAY}"
```
Comment (stas-at-ibm) [inline docker-compose.yaml:18]: Update README setup portion with the new env variable.

**Review by stas-at-ibm — CHANGES_REQUESTED** [review]
Nice one bro! Check my review below!

File: backend/src/utils/config.loader.js (hunk 4 above)
Comment (lukatarman) [reply backend/src/utils/config.loader.js]: Hey when you review can you leave me a message on discord explaining why we do this? Why put this into another sub property called options? And how come "expectedErrorTypes" isn't under options in runner? Just wanna understand it :)

**Review by stas-at-ibm — CHANGES_REQUESTED** [review]
Great job as always man! See my review below.

File: backend/src/utils/time.utils.spec.js
Comment (stas-at-ibm) [inline backend/src/utils/time.utils.spec.js:1]: Just beautiful! Thank you so much for those awesome tests :)

Now you know more and more your way around a complex API documentation like jasmine ;)...its a smal library but so powerful for testing! It has everything if you're able to find and use it

File: backend/src/utils/time.utils.spec.js [hunk 7]
```diff
@@ -8,47 +8,60 @@ import {
 
 describe("time.utils.js", () => {
   describe(".delay", () => {
-    let timePassed;
+    describe("when we wait for a delay of 0", () => {
+      beforeEach(async function () {
+        jasmine.clock().install();
+        jasmine.clock().mockDate(new Date("September 2020"));
```
Comment (stas-at-ibm) [inline backend/src/utils/time.utils.spec.js:14]: Do you have to set a mock date for it to work?

File: backend/src/utils/runner.spec.js [hunk 8]
```diff
@@ -57,44 +57,81 @@ describe("runner.js", () => {
...
-        it("and calls the function twice", async () => {
-          const funcSpy = jasmine.createSpy("funcSpy");
-          const result = new Runner(
-            createLoggerMock(),
-            options(delayMock, 0, 2),
-          ).runAsync([funcSpy], []);
+      describe("for two iterations", () => {
+        describe("with a set delay", () => {
+          beforeEach(async function () {
+            this.funcOneSpy = jasmine.createSpy("funcOneSpy");
```
Comment (stas-at-ibm) [inline backend/src/utils/runner.spec.js:63]: `funcSpy`

File: backend/src/utils/runner.spec.js [hunk 9]
```diff
@@ -57,44 +57,81 @@ describe("runner.js", () => {
...
+            this.funcTwoSpy = jasmine.createSpy("funcTwoSpy");
```
Comment (stas-at-ibm) [inline backend/src/utils/runner.spec.js:64]: `delaySpy`

File: backend/src/utils/runner.spec.js [hunk 10]
```diff
@@ -57,44 +57,81 @@ describe("runner.js", () => {
...
+
+          it("the function ran two times", async function () {
+            await expectAsync(this.runnerResult).toBeResolved();
+            expect(this.funcOneSpy).toHaveBeenCalledTimes(2);
+          });
+        });
+
+        describe("with the default delay", () => {
+          beforeEach(async function () {
+            this.funcOneSpy = jasmine.createSpy("funcOneSpy");
```
Comment (stas-at-ibm) [inline backend/src/utils/runner.spec.js:88]: `funcSpy`

File: backend/src/utils/runner.spec.js [hunk 11]
```diff
@@ -57,44 +57,81 @@ describe("runner.js", () => {
...
+            this.funcTwoSpy = jasmine.createSpy("funcTwoSpy");
```
Comment (stas-at-ibm) [inline backend/src/utils/runner.spec.js:89]: `delaySpy`

File: backend/src/utils/runner.spec.js [hunk 12]
```diff
@@ -57,44 +57,81 @@ describe("runner.js", () => {
...
+
+            this.runnerResult = runner.runAsync(
+              [{ func: this.funcOneSpy, delay: 2000 }],
+              [],
+            );
+
+            await this.runnerResult;
+          });
+
+          it("the function was called with the correct argument", async function () {
```
Comment (stas-at-ibm) [inline backend/src/utils/runner.spec.js:76]: remove async

File: backend/src/utils/runner.spec.js [hunk 13]
```diff
@@ -57,44 +57,81 @@ describe("runner.js", () => {
...
+            this.funcTwoSpy = jasmine.createSpy("funcTwoSpy");
+
+            const runner = new Runner(createLoggerMock(), options(this.funcTwoSpy, 2));
+
+            this.runnerResult = runner.runAsync([{ func: this.funcOneSpy }], []);
+
+            await this.runnerResult;
+          });
+
+          it("the function was called with the correct argument", async function () {
```
Comment (stas-at-ibm) [inline backend/src/utils/runner.spec.js:98]: remove async

File: backend/src/utils/config.loader.mock.js
Comment (stas-at-ibm) [inline backend/src/utils/config.loader.mock.js:1]: keep the structure of the config object, but no need to set them all, only set what you need as you did

File: backend/src/features/player-history-aggregator/player.history.aggregator.spec.js [hunk 14]
```diff
@@ -230,7 +237,7 @@ describe("PlayerHistoryAggregator", function () {
   });
 
   describe(".addCurrentPlayers()", () => {
-    describe("does not have any games to check for current player numbers", function () {
+    fdescribe("does not have any games to check for current player numbers", function () {
```
Comment (stas-at-ibm) [inline backend/src/features/player-history-aggregator/player.history.aggregator.spec.js:240]: remove f

File: backend/src/features/player-history-aggregator/player.history.aggregator.spec.js [hunk 15]
```diff
@@ -183,7 +184,7 @@ describe("PlayerHistoryAggregator", function () {
       });
     });
 
-    describe("finds no games in the database and finishes", () => {
+    fdescribe("finds no games in the database and finishes", () => {
```
Comment (stas-at-ibm) [inline backend/src/features/player-history-aggregator/player.history.aggregator.spec.js]: remove f

File: backend/src/utils/config.loader.mock.js [hunk 16]
```diff
@@ -0,0 +1,13 @@
+export function configOptions(
```
Comment (stas-at-ibm) [inline backend/src/utils/config.loader.mock.js:1]: rename to `createConfigMock`...its a factory functions...good practice to use create in the name if something is created

**Review by stas-at-ibm — APPROVED** [review]
Grat work my friend! See my comments below!

**Review by stas-at-ibm — CHANGES_REQUESTED** [review]
Sorry, clicked the wrong button! :)

File: backend/src/features/game-identifier/game.identifier.spec.js [hunk 17]
```diff
@@ -16,12 +16,17 @@ import { createLoggerMock } from "../../utils/logger.mock.js";
...
 import { getXGamesWithoutDetails } from "../../models/game.mocks.js";
-import { configOptions } from "../../utils/config.loader.mock.js";
+import { createConfigMock } from "../../utils/config.loader.mock.js";
 
 describe("game.identifier.js", function () {
   describe(".tryViaSteamWeb", function () {
     describe("gets zero steamApps from the database and stops. So,", function () {
-      beforeEach(async function () {
+      beforeAll(async function () {
+        this.options = {
```
Comment (stas-at-ibm) [inline backend/src/features/game-identifier/game.identifier.spec.js:25]: This one sneaked in, you can remove it.

Remove everywhere.

**Review by stas-at-ibm — APPROVED** [review]
Good job! Just remove the options and we can merge after we squashed and rebased.

**Review by stas-at-ibm — APPROVED** [review]
Great! Lets merge!

**Pull request merged**